---
title: "plant-food-research-open/genepal Report"
author: "`r Sys.getenv('USER')`"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---

```{r setup, warning=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE)

suppressPackageStartupMessages({
  library(viridis)
  library(tidyr)
  library(dplyr)
  library(stringr)
  library(pheatmap)
  library(kableExtra)
  library(gt)
  library(gtExtras)
  library(tibble)
  library(jsonlite)
  library(yaml)
})

source("./genepal_report.R")

outputs_folder <- "genepal_data"
if (!dir.exists(outputs_folder)) {
  dir.create(outputs_folder, recursive = TRUE)
}

report_data_json <- list()
```


```{css}
.main-container {
max-width: 95% !important;
margin: auto;
}

table.gt_table {
margin-left: 0 !important;
width: 100% !important;
}

body {
padding-left: 0 !important;
margin-left: 0 !important;
width: 100% !important;
}

.caption {
text-align: left;
}

.wf_summary_container {
word-wrap: break-word;
}
```


## General Statistics

```{r genes_features_annotation, warning=FALSE}
gff_combined_results <-
  list.files("results/etc/splicing_marked/",
    pattern = "\\.gff3$", full.names = TRUE
  ) %>%
  lapply(summarise_gff) %>%
  bind_rows() %>%
  (function(df) {
    df[is.na(df)] <- 0
    df
  }) %>%
  dplyr::select(ID, gene, mRNA, CDS, exon, intron, Non_canon_splice_sites) %>%
  (function(df) {
    colnames(df) <- c("ID", "Genes", "mRNA", "CDS", "Exons", "Intron", "Non canon splice sites")
    df
  })

egg_plus_gff <- merge(gff_combined_results, extract_eggnog_stats())

if (nrow(egg_plus_gff) < 1) {
  genes_and_annos_table <- gff_combined_results

  genes_and_annos_table_gt <-
    genes_and_annos_table %>%
    gt() %>%
    tab_header(title = md("**Gene prediction**")) %>%
    tab_options(table.align = "left", container.width = pct(60), table.width = pct(60), table.margin.left = px(0), table.margin.right = px(0)) %>%
    tab_style(
      style = cell_text(align = "left"),
      locations = cells_title(groups = "title")
    )
  write.csv(x = gff_combined_results, file = "./genes.anno.report.csv", row.names = FALSE)
} else {
  genes_and_annos_table <- egg_plus_gff

  genes_and_annos_table_gt <-
    genes_and_annos_table %>%
    gt() %>%
    tab_header(title = md("**Gene prediction and annotation**")) %>%
    tab_options(table.align = "left", container.width = pct(50), table.width = pct(50), table.margin.left = px(0), table.margin.right = px(0)) %>%
    tab_style(
      style = cell_text(align = "left"),
      locations = cells_title(groups = "title")
    )
  write.csv(x = egg_plus_gff, file = file.path(outputs_folder, "genes.anno.report.csv"), row.names = FALSE)
}

report_data_json <- append(report_data_json, list(stats = genes_and_annos_table))

genes_and_annos_table_gt
```


<!-- BUSCO -->


```{r busco_merged_final_df, warning=FALSE}
busco_merged_final_df <-
  parse_busco_folder("results/busco/fasta/", col_prefix = "Genome") %>%
  merge(
    parse_busco_folder("results/busco/gff/", col_prefix = "Gene")
  )
```


```{r busco_results, eval=(exists("busco_merged_final_df") && nrow(busco_merged_final_df) > 0), results='asis', warning=FALSE}
cat("## BUSCO")

busco_merged_final_df %>%
  select(ID, Lineage, 3, 8, 4, 9, 5, 10, 6, 11, 7, 12) %>%
  gt() %>%
  tab_header(title = md("**BUSCO completeness estimation**")) %>%
  tab_options(table.align = "left", container.width = pct(100), table.width = pct(100), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )

write.csv(x = busco_merged_final_df, file = file.path(outputs_folder, "busco.report.csv"), row.names = FALSE)

report_data_json <- append(report_data_json, list(busco = busco_merged_final_df))
```


<!-- OrthoFinder Statistics -->


```{r final_orthstats, warning=FALSE}
oof <- "results/orthofinder/genepal/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv"
final_orthstats <- process_protein_clustering(oof)
```


```{r orthofinder_results, eval=(exists("final_orthstats") && !is.null(final_orthstats)), results='asis', warning=FALSE}
cat("## OrthoFinder")

final_orthstats %>%
  gt() %>%
  tab_header(title = md("**Gene clustering statistics**")) %>%
  tab_options(table.align = "left", container.width = pct(100), table.width = pct(100), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )
```


<!-- OrthoFinder Hierarchical Orthogroups -->


```{r n0_df, warning=FALSE}
n0o <- "results/orthofinder/genepal/Phylogenetic_Hierarchical_Orthogroups/N0.tsv"
n0_df <- process_hog_counts(n0o)
```


```{r pangenome_matrix_heading, eval=(exists("n0_df") && !is.null(n0_df$heatmap)), results='asis', warning=FALSE}
cat("### Pangenome Matrix")
cat("<br>")
cat("<br>")
```


```{r pheatmap, eval=(exists("n0_df") && !is.null(n0_df$heatmap)), results='hide', fig.align='center', fig.cap="Heatmap showing number of proteins present in each orthocluster (clusters where all individuals have 1 copy are excluded). Columns = Orthologue cluster, Row = Individual", fig.width=7, fig.height=7, dpi=150, warning=FALSE}

# Max 65536 allowed
# https://github.com/Plant-Food-Research-Open/genepal/issues/124

n_cols <- ncol(n0_df$heatmap)
max_cols_allowed <- min(n_cols, 5000)

# Approach 1: Random selection of columns
# selected_cols <- sample(n_cols, max_cols_allowed)

# Approach 2: First N largest clusters
selected_cols <- order(colSums(n0_df$heatmap), decreasing = TRUE)[seq(1, max_cols_allowed)]

prefix_text <- ""

if ( n_cols != max_cols_allowed ) {
  prefix_text <- paste0("Top ", max_cols_allowed, " ")
}

pheatmap(n0_df$heatmap[, selected_cols],
  show_colnames = FALSE,
  main = paste0(prefix_text, "Orthologue clusters"),
  legend = TRUE,
  legend_labels = TRUE,
  border_color = "white"
)

pheatmap(n0_df$heatmap[, selected_cols],
  filename = file.path(outputs_folder, "pangene.matrix.heatmap.pdf"),
  show_colnames = FALSE,
  main = paste0(prefix_text, "Orthologue clusters"),
  legend = TRUE,
  legend_labels = TRUE,
  border_color = "white"
)

write.csv(x = transform_hogs(n0o), file = file.path(outputs_folder, "pangenome.matrix.csv"), row.names = FALSE)
```


<!-- Benchmarking Statistics -->


```{r benchmarking_statistics, warning=FALSE}
benchmarking_statistics <- parse_gff_compare_folder("results/benchmark/")
```


```{r eval=(exists("benchmarking_statistics") && !is.null(benchmarking_statistics)), results='asis', warning=FALSE}
cat("## Benchmarking Statistics")

benchmarking_statistics$sensitivity_precision %>%
  gt() %>%
  tab_options(table.align = "left", container.width = pct(50), table.width = pct(50), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_header(title = md("**Benchmarking: sensitivity and precision**")) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )

cat("<br>")

benchmarking_statistics$matching %>%
  gt() %>%
  tab_options(table.align = "left", container.width = pct(50), table.width = pct(50), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_header(title = md("**Benchmarking: matching introns, transcripts, and loci**")) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )

cat("<br>")

benchmarking_statistics$missing_novel %>%
  gt() %>%
  tab_options(table.align = "left", container.width = pct(50), table.width = pct(50), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_header(title = md("**Benchmarking: missing and novel features**")) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )

report_data_json <- append(report_data_json, list(benchmark = benchmarking_statistics))
```

```{r store_json, warning=FALSE}
toJSON(report_data_json, pretty = TRUE) %>%
  write(file.path(outputs_folder, "summary_stats.json"))
```

```{r print_parameters, eval=(file.exists("results/pipeline_info/workflow_summary_mqc.yaml")), results='asis', warning=FALSE}
cat("<h2> Workflow Summary </h2>")

cat('<div class="wf_summary_container">')
cat(read_yaml("results/pipeline_info/workflow_summary_mqc.yaml")$data)
cat("</div>")
```

```{r print_versions, eval=(file.exists("results/pipeline_info/genepal_software_mqc_versions.yml")), results='asis', warning=FALSE}
cat("<h2> Software Versions </h2>")

versions_df <-
  read_yaml("results/pipeline_info/genepal_software_mqc_versions.yml") %>%
  unlist() %>%
  data.frame()

colnames(versions_df) <- "Version"

versions_df <-
  versions_df %>%
  mutate(
    Group = sapply(str_split(rownames(versions_df), pattern = "\\."), "[", 1),
    Software = sapply(str_split(rownames(versions_df), pattern = "\\."), "[", 2)
  ) %>%
  select(Group, Software, Version)

rownames(versions_df) <- NULL

versions_df %>%
  gt() %>%
  tab_options(table.align = "left", container.width = pct(70), table.width = pct(70), table.margin.left = px(0), table.margin.right = px(0)) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "title")
  )
```

```{r print_wf_description, eval=(file.exists("results/pipeline_info/methods_description_mqc.yaml")), results='asis', warning=FALSE}
cat("<h2> Methods Description </h2>")

cat('<div class="wf_summary_container">')
cat(read_yaml("results/pipeline_info/methods_description_mqc.yaml")$data)
cat("</div>")
```
