# plant-food-research-open/genepal: Citations

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- AGAT, [GPL v3](https://github.com/NBISweden/AGAT/blob/master/LICENSE)

  > Jacques Dainat, Darío Hereñú, Dr. K. D. Murray, Ed Davis, Ivan Ugrin, Kathryn Crouch, LucileSol, Nuno Agostinho, pascal-git, Zachary Zollman, & tayyrov. (2024). NBISweden/AGAT: AGAT-v1.4.1 (v1.4.1). Zenodo. doi: <a href="https://doi.org/10.5281/zenodo.13799920">10.5281/zenodo.13799920</a>

- AUGUSTUS, [Artistic license-1.0](https://github.com/Gaius-Augustus/Augustus/blob/master/src/LICENSE.TXT)

  > Sommerfeld, D., Lingner, T., Stanke, M., Morgenstern, B., & Richter, H. (2009). AUGUSTUS at MediGRID: Adaption of a bioinformatics application to grid computing for efficient genome analysis. Future Gener. Comput. Syst., 25, 337-345. doi: <a href="https://doi.org/10.1016/j.future.2008.05.010">10.1016/j.future.2008.05.010</a>

- BRAKER3, [Artistic license-1.0](https://github.com/Gaius-Augustus/BRAKER/blob/master/LICENSE.TXT)

  > Gabriel, L., Bruna, T., Hoff, K. J., Ebel, M., Lomsadze, A., Borodovsky, M., Stanke, M. (2023). BRAKER3: Fully Automated Genome Annotation Using RNA-Seq and Protein Evidence with GeneMark-ETP, AUGUSTUS and TSEBRA. bioRxiV, doi: <a href="https://doi.org/10.1101/2023.06.10.544449">10.1101/2023.06.10.544449</a>

- BUSCO, [MIT](https://gitlab.com/ezlab/busco/-/blob/master/LICENSE)

  > Manni M, Berkeley MR, Seppey M, Simão FA, Zdobnov EM. 2021. BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes, Molecular Biology and Evolution, Volume 38, Issue 10, October 2021, Pages 4647–4654, doi: <a href="https://doi.org/10.1093/molbev/msab199">10.1093/molbev/msab199</a>

- EDTA, [GPL v3](https://github.com/oushujun/EDTA/blob/master/LICENSE)

  > Ou S., Su W., Liao Y., Chougule K., Agda J. R. A., Hellinga A. J., Lugo C. S. B., Elliott T. A., Ware D., Peterson T., Jiang N., Hirsch C. N. and Hufford M. B. (2019). Benchmarking Transposable Element Annotation Methods for Creation of a Streamlined, Comprehensive Pipeline. Genome Biol. 20(1): 275. doi: <a href="https://doi.org/10.1186/s13059-019-1905-y">10.1186/s13059-019-1905-y</a>

- EggNOG-mapper, [GPL v3](https://github.com/eggnogdb/eggnog-mapper/blob/master/LICENSE.txt)

  > eggNOG-mapper v2: functional annotation, orthology assignments, and domain prediction at the metagenomic scale. Carlos P. Cantalapiedra, Ana Hernandez-Plaza, Ivica Letunic, Peer Bork, Jaime Huerta-Cepas. 2021. Molecular Biology and Evolution, msab293, doi: <a href="https://doi.org/10.1093/molbev/msab293">10.1093/molbev/msab293</a>

- FASTQC, [GPL v3](https://github.com/s-andrews/FastQC/blob/master/LICENSE.txt)

  > Andrews, S. (2010). Babraham Bioinformatics - FastQC a quality control tool for high throughput sequence data. Babraham.ac.uk. url: <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc">https://www.bioinformatics.babraham.ac.uk/projects/fastqc</a>

- FASTP, [MIT](https://github.com/OpenGene/fastp/blob/master/LICENSE)

  > Chen S, Zhou Y, Chen Y, Gu J. 2018. fastp: an ultra-fast all-in-one FASTQ preprocessor, Bioinformatics, Volume 34, Issue 17, 01 September 2018, Pages i884–i890, doi: <a href="https://doi.org/10.1093/bioinformatics/bty560">10.1093/bioinformatics/bty560</a>

- GeneMark-ETP, [Attribution-NonCommercial-ShareAlike 4.0 International](https://github.com/gatech-genemark/GeneMark-ETP/blob/main/License-Creative-Commons-Attribution-NonCommercial-ShareAlike-4.0-International.txt)

  > Brůna T, Lomsadze A, Borodovsky M. GeneMark-ETP significantly improves the accuracy of automatic annotation of large eukaryotic genomes. Genome Res. 2024 Jun 25;34(5):757-768. doi: <a href="https://doi.org/10.1101/gr.278373.123">10.1101/gr.278373.123</a>. PMID: 38866548; PMCID: PMC11216313.

- GenomeTools, [ISC](http://genometools.org/license.html)

  > Gremme G, Steinbiss S, Kurtz S. 2013. "GenomeTools: A Comprehensive Software Library for Efficient Processing of Structured Genome Annotations," in IEEE/ACM Transactions on Computational Biology and Bioinformatics, vol. 10, no. 3, pp. 645-656, May 2013, doi: <a href="https://doi.org/10.1109/TCBB.2013.68">10.1109/TCBB.2013.68</a>

- GffCompare, [MIT](https://github.com/gpertea/gffcompare/blob/master/LICENSE)

  > Pertea G, Pertea M. GFF Utilities: GffRead and GffCompare. F1000Res. 2020 Apr 28;9:ISCB Comm J-304. doi: <a href="http://doi.org/10.12688/f1000research.23297.2">10.12688/f1000research.23297.2</a>. PMID: 32489650; PMCID: PMC7222033.

- GffRead, [MIT](https://github.com/gpertea/gffread/blob/master/LICENSE)

  > Pertea G, Pertea M. GFF Utilities: GffRead and GffCompare. F1000Res. 2020 Apr 28;9:ISCB Comm J-304. doi: <a href="http://doi.org/10.12688/f1000research.23297.2">10.12688/f1000research.23297.2</a>. PMID: 32489650; PMCID: PMC7222033.

- Liftoff, [GPL v3](https://github.com/agshumate/Liftoff/blob/master/LICENSE.md)

  > Shumate A, Salzberg SL. Liftoff: accurate mapping of gene annotations. Bioinformatics. 2021 Jul 19;37(12):1639-1643. doi: <a href="http://doi.org/10.1093/bioinformatics/btaa1016">10.1093/bioinformatics/btaa1016</a>. PMID: 33320174; PMCID: PMC8289374.

- OrthoFinder, [GPL v3](https://github.com/davidemms/OrthoFinder/blob/master/License.md)

  > Emms, D.M., Kelly, S. OrthoFinder: phylogenetic orthology inference for comparative genomics. Genome Biol 20, 238 (2019). doi: <a href="https://doi.org/10.1186/s13059-019-1832-y">10.1186/s13059-019-1832-y</a>

- ProtHint, [License for GeneMark family software ("Product")](https://github.com/gatech-genemark/ProtHint/blob/master/LICENSE)

  > Tomáš Brůna, Alexandre Lomsadze, Mark Borodovsky, GeneMark-EP+: eukaryotic gene prediction with self-training in the space of genes and proteins, NAR Genomics and Bioinformatics, Volume 2, Issue 2, June 2020, lqaa026, doi: <a href="https://doi.org/10.1093/nargab/lqaa026">10.1093/nargab/lqaa026</a>

- py_fasta_validator, [MIT](https://github.com/linsalrob/py_fasta_validator/blob/master/LICENSE)

  > Edwards, R.A. 2019. fasta_validate: a fast and efficient fasta validator written in pure C. doi: <a href="https://doi.org/10.5281/zenodo.2532044">10.5281/zenodo.2532044</a>

- RepeatMasker, [Open Software License v. 2.1](https://github.com/rmhubley/RepeatMasker/blob/master/LICENSE)

  > Smit, A., & Hubley, R. (2023). RepeatMasker: a program that screens DNA sequences for interspersed repeats and low complexity DNA sequences. Repeatmasker.org. url: <a href="https://www.repeatmasker.org">https://www.repeatmasker.org</a>

- RepeatModeler, [Open Software License v. 2.1](https://github.com/Dfam-consortium/RepeatModeler/blob/master/LICENSE)

  > Hubley, R. (2023). RepeatModeler: a de novo transposable element (TE) family identification and modeling package. Repeatmasker.org. url: <a href="https://www.repeatmasker.org">https://www.repeatmasker.org</a>

- Samtools, [MIT/Expat](https://github.com/samtools/samtools/blob/develop/LICENSE)

  > Danecek P, Bonfield JK, Liddle J, Marshall J, Ohan V, Pollard MO, Whitwham A, Keane T, McCarthy SA, Davies RM, Li H. 2021. Twelve years of SAMtools and BCFtools, GigaScience, Volume 10, Issue 2, February 2021, giab008, doi: <a href="https://doi.org/10.1093/gigascience/giab008">10.1093/gigascience/giab008</a>

- SeqKit, [MIT](https://github.com/shenwei356/seqkit/blob/master/LICENSE)

  > Shen W, Le S, Li Y, Hu F. 2016. SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation. PLoS ONE 11(10): e0163962. doi: <a href="https://doi.org/10.1371/journal.pone.0163962">10.1371/journal.pone.0163962</a>

- SortMeRNA, [GPL v3](https://github.com/sortmerna/sortmerna/blob/master/LICENSE.txt)

  > Kopylova E., Noé L. and Touzet H., "SortMeRNA: Fast and accurate filtering of ribosomal RNAs in metatranscriptomic data", Bioinformatics (2012), doi: <a href="https://doi.org/10.1093/bioinformatics/bts611">10.1093/bioinformatics/bts611</a>

- STAR, [MIT](https://github.com/alexdobin/STAR/blob/master/LICENSE)

  > Dobin A, Davis CA, Schlesinger F, Drenkow J, Zaleski C, Jha S, Batut P, Chaisson M, Gingeras TR. STAR: ultrafast universal RNA-seq aligner. Bioinformatics. 2013 Jan 1;29(1):15-21. doi: <a href="https://doi.org/10.1093/bioinformatics/bts635">10.1093/bioinformatics/bts635</a>. Epub 2012 Oct 25. PMID: 23104886; PMCID: PMC3530905.

- TSEBRA, [The Artistic License 2.0](https://github.com/Gaius-Augustus/TSEBRA/blob/main/bin/LICENSE.txt)

  > Gabriel, L., Hoff, K.J., Brůna, T. et al. TSEBRA: transcript selector for BRAKER. BMC Bioinformatics 22, 566 (2021). doi: <a href="https://doi.org/10.1186/s12859-021-04482-0">10.1186/s12859-021-04482-0</a>

## Software packaging/containerisation tools

- [Anaconda](https://anaconda.com)

  > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

- [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

  > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

- [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

  > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

- [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

  > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

- [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

  > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
