# plant-food-research-open/genepal: Output

## Introduction

This document describes the output produced by the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

<!-- no toc -->

- [Repeat annotation](#repeat-annotation)
- [Repeat masking](#repeat-masking)
- [RNAseq trimming, filtering and QC](#rnaseq-trimming-filtering-and-qc)
- [RNAseq Alignment](#rnaseq-alignment)

### Repeat annotation

<details markdown="1">
<summary>Output files</summary>

- `repeatmodeler/`
  - `*.fa`: Repeat library
- `edta/`
  - `*.EDTA.TElib.fa`: Repeat library

</details>

A repeat library is created with either [REPEATMODELER](https://github.com/Dfam-consortium/RepeatModeler) or [EDTA](https://github.com/oushujun/EDTA). The choice of the tool is specified by the `repeat_annotator` parameter (default: `repeatmodeler`). Repeat annotation outputs are saved to the output directory only if `save_annotated_te_lib` parameter is set to `true` (default: `false`).

### Repeat masking

<details markdown="1">
<summary>Output files</summary>

- `repeatmasker/`
  - `*.masked`: Masked assembly
  - `*.gff`: Repeatmasker output in GFF3 format

</details>

Soft masking of the repeats is performed with [REPEATMASKER](https://github.com/rmhubley/RepeatMasker) using the repeat library prepared in the previous step. Masking outputs are saved to the output directory only if `repeatmasker_save_outputs` parameter is set to `true` (default: `false`).

### RNAseq trimming, filtering and QC

<details markdown="1">
<summary>Output files</summary>

- `fastqc_raw/`
  - `*.html`: HTML QC report for a sample before trimming
  - `*.zip`: Zipped QC files for a sample before trimming
- `fastqc_trim/`
  - `*.html`: HTML QC report for a sample after trimming
  - `*.zip`: Zipped QC files for a sample after trimming
- `fastp/`
  - `html/`
    - `*.fastp.html`: HTML trimming report for a sample
  - `json/`
    - `*.fastp.json`: Trimming statistics for a sample
  - `log/`
    - `*.fastp.log`: Trimming log for a sample
  - `*_{1,2}.fail.fastq.gz`: Reads which failed trimming
  - `*.paired.fail.fastq.gz`: Pairs of reads which failed trimming
  - `*.merged.fastq.gz`: Reads which passed trimming. For paired reads, reads 1 and 2 are merged into a single file
- `sortmerna/`
  - `*.sortmerna.log`: Filtering log for a sample
  - `*_{1,2}.non_rRNA.fastq.gz`: Filtered reads

</details>

RNASeq reads are trimmed with [FASTP](https://github.com/OpenGene/fastp) and are QC'ed with [FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc). Ribosomal reads are filtered out using [SORTMERNA](https://github.com/sortmerna/sortmerna). Trimmed reads are only stored to the output directory if the `save_trimmed` parameter is set to `true` (default: `false`). Reads filtered by [SORTMERNA](https://github.com/sortmerna/sortmerna) are stored to the output directory if the `save_non_ribo_reads` parameter is set to `true` (default: `false`).

### RNAseq Alignment

<details markdown="1">
<summary>Output files</summary>

- `star/`
  - `alignment/`
    - `X.on.Y.Aligned.sortedByCoord.out.bam`: Sorted BAM file of read alignments for sample `X` against reference `Y`
    - `X.on.Y.Log.final.out`: STAR final log file for sample `X` against reference `Y`
  - `cat_bam/`
    - `Y.bam`: A single BAM file for reference `Y` created by concatenating alignments from sample-wise `*.on.Y.Aligned.sortedByCoord.out.bam` files

</details>

RNAseq alignment is performed with [STAR](https://github.com/alexdobin/STAR). Alignment files are only stored to the output directory if the `star_save_outputs` parameter is set to `true` (default: `false`). Concatenated bam files are stored to the output directory if the `save_cat_bam` parameter is set to `true` (default: `false`).

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
